### PR TITLE
[export] make aot_export_module uses dynamo's fake_mode

### LINKED ISSUE #110100: torch.export.export does not support already-fakefied inputs within FakeTensorMode
### 🐛 Describe the bug

Apparently `torch.export.export` does not support already fakefied inputs that `torch._dynamo.export` does

```python
import torch
from torch._subclasses import fake_tensor

fake_mode = fake_tensor.FakeTensorMode()

class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.linear = torch.nn.Linear(2, 2)

    def forward(self, x):
        out = self.linear(x)
        return out

with fake_mode:
    x = torch.rand(5, 2, 2)
    model = Model()
    # gm, _ = torch._dynamo.export(model, x)  # this works
    exported_program = torch.export.export(model, (x,))  # this fails with AssertionError: fake mode (<torch._subclasses.fake_tensor.FakeTensorMode object at 0x7fe4259ac760>) from active fake mode 0 doesn't match mode (<torch._subclasses.fake_tensor.FakeTensorMode object at 0x7fe355ae0760>) from fake tensor input 0

```

The error is

```bash
[2023-09-26 19:46:35,122] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/opt/conda/envs/ptca/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_validation.py:114: UserWarning: WARNING: failed to get cudart_version from onnxruntime build info.
  warnings.warn("WARNING: failed to get cudart_version from onnxruntime build info.")
Traceback (most recent call last):
  File "repro_fake_tensor_exception_with_torch_export_export.py", line 19, in <module>
    gm, _ = torch.export.export(model, x)
  File "/opt/pytorch/torch/export/__init__.py", line 587, in export
    return export__RC__(f, args, kwargs, dynamic_shapes=dynamic_shapes)
  File "/opt/pytorch/torch/_export/__init__.py", line 92, in export__RC__
    return export(f, args, kwargs)
  File "/opt/pytorch/torch/_export/__init__.py", line 395, in export
    raise UserError(UserErrorType.INVALID_INPUT,
torch._dynamo.exc.UserError: Expecting `args` to be a tuple of example positional inputs, got <class 'torch._subclasses.fake_tensor.FakeTensor'>
(ptca) root@88c3e49b4bcf:/opt/pytorch# python repro_fake_tensor_exception_with_torch_export_export.py 
[2023-09-26 19:46:56,376] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/opt/conda/envs/ptca/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_validation.py:114: UserWarning: WARNING: failed to get cudart_version from onnxruntime build info.
  warnings.warn("WARNING: failed to get cudart_version from onnxruntime build info.")
> /opt/pytorch/torch/_export/__init__.py(430)export()
-> fake_args, fake_kwargs, fake_mode = _convert_input_to_fake(gm_torch_level, args, kwargs)
(Pdb) c
Traceback (most recent call last):
  File "repro_fake_tensor_exception_with_torch_export_export.py", line 19, in <module>
    gm, _ = torch.export.export(model, (x,))
  File "/opt/pytorch/torch/export/__init__.py", line 587, in export
    return export__RC__(f, args, kwargs, dynamic_shapes=dynamic_shapes)
  File "/opt/pytorch/torch/_export/__init__.py", line 92, in export__RC__
    return export(f, args, kwargs)
  File "/opt/pytorch/torch/_export/__init__.py", line 430, in export
    fake_args, fake_kwargs, fake_mode = _convert_input_to_fake(gm_torch_level, args, kwargs)
  File "/opt/pytorch/torch/_export/__init__.py", line 289, in _convert_input_to_fake
    if detected_fake_mode := detect_fake_mode(fake_inps):
  File "/opt/pytorch/torch/_guards.py", line 829, in detect_fake_mode
    assert fake_mode is m, (
AssertionError: fake mode (<torch._subclasses.fake_tensor.FakeTensorMode object at 0x7fe4259ac760>) from active fake mode 0 doesn't match mode (<torch._subclasses.fake_tensor.FakeTensorMode object at 0x7fe355ae0760>) from fake tensor input 0

fake mode from active fake mode 0 allocated at:
  File "repro_fake_tensor_exception_with_torch_export_export.py", line 4, in <module>
    fake_mode = fake_tensor.FakeTensorMode()
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1273, in __init__
    self.stack = "".join(traceback.format_stack())

fake mode from fake tensor input 0 allocated at:
  File "repro_fake_tensor_exception_with_torch_export_export.py", line 19, in <module>
    gm, _ = torch.export.export(model, (x,))
  File "/opt/pytorch/torch/export/__init__.py", line 587, in export
    return export__RC__(f, args, kwargs, dynamic_shapes=dynamic_shapes)
  File "/opt/pytorch/torch/_export/__init__.py", line 92, in export__RC__
    return export(f, args, kwargs)
  File "/opt/pytorch/torch/_export/__init__.py", line 407, in export
    gm_torch_level, _ = torch._dynamo.export(
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1216, in inner
    result_traced = opt_f(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1519, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1528, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1519, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1528, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/opt/pytorch/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/opt/pytorch/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/opt/pytorch/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/opt/pytorch/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/opt/pytorch/torch/_dynamo/convert_frame.py", line 434, in transform
    tracer = InstructionTranslator(
  File "/opt/pytorch/torch/_dynamo/symbolic_convert.py", line 2007, in __init__
    output=OutputGraph(
  File "/opt/pytorch/torch/_dynamo/output_graph.py", line 291, in __init__
    fake_mode = torch._subclasses.FakeTensorMode(
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1273, in __init__
    self.stack = "".join(traceback.format_stack())

(ptca) root@88c3e49b4bcf:/opt/pytorch# 
```

### Versions

pytorch main branch

cc @eellison @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113681

Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module even on a machine that installs a cpu version of pytorch. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo

Differential Revision: [D51414973](https://our.internmc.facebook.com/intern/diff/D51414973)

### CODE REVIEW

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -563,7 +563,17 @@ def export(
         preserve_module_call_signature=preserve_module_call_signature,
     )
 
+def disable_prexisiting_fake_mode(fn):
```
Comment (zhxchen17) [inline torch/_export/__init__.py:566]:
```suggestion
def _disable_prexisiting_fake_mode(fn):
```

**Review by zhxchen17 — COMMENTED** [review]
Nice! Could you also link the issue number in the PR summary so we can get a reference there?

File: torch/_export/__init__.py (hunk 1 above)
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:566]: Does this mean we are always using dynamo produced fake mode?

File: torch/_export/__init__.py [hunk 2]
```diff
@@ -695,11 +706,12 @@ def _export(
 
     # Note: aot_export_module doesn't accept kwargs, we'd like to reorder the kwargs as an OrderedDict
     # to follow the order in orig_args and correctly call gm_torch_level
-    gm, graph_signature = aot_export_module(
-        gm_torch_level,
-        (*fake_args, *_reorder_kwargs_by_names(orig_args, fake_args, fake_kwargs).values()),
-        trace_joint=False
-    )
+    with torch.nn.utils.stateless._reparametrize_module(gm_torch_level, fake_params_buffers):
```
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:709]: Just curious, why is this needed?

File: torch/_export/__init__.py (hunk 1 above)
Comment (ydwu4) [reply torch/_export/__init__.py:566]: yes! dynamo's fake_mode will be the single source of truth. This is easier to reason about and it also makes sure we won't mutate user provided fake_mode. The user provided input and model.named_parameters and buffers are still using outer fake_mode.

File: torch/_export/__init__.py (hunk 2 above)
Comment (ydwu4) [reply torch/_export/__init__.py:709]: The gm_torch_level.named_parameters().fake_mode is the outer fake_mode because dynamo doesn't wrap them as fake tensors (unlike what it does for inputs) and register the original nn_modules as it is to gm_torch_level. 

This reparametrize_module make sure inputs and gm_torch_level.params/buffers have the same fake_mode, otherwise aot_export_module will error out because it sees a mix of fake_modes.

Comment (SherlockNoMad) [reply torch/_export/__init__.py:709]: >This reparametrize_module makes sure inputs and gm_torch_level.params/buffers have the same fake_mode, otherwise aot_export_module will error out because it sees a mix of fake_modes. And we want aot_export_module to use the fake_tensor mode in dynamo to keep the pipeline easy to reason about.

nit, add this as comment.

File: test/export/test_export.py [hunk 3]
```diff
@@ -1548,6 +1548,75 @@ def forward(self, x):
...
+            def forward(self, x):
+                out = self.linear(x)
+                return out
+
+        # Put the inputs on a device
+        with fake_mode, torch.device('meta'):
+            x = torch.rand(5, 2, 2)
+            model = Model()
+
+        # Mannuly se the fake_device of fake tensors.
```
Comment (SherlockNoMad) [inline test/export/test_export.py:1596]: set

File: test/export/test_export.py [hunk 4]
```diff
@@ -1548,6 +1548,75 @@ def forward(self, x):
...
+        for n, p in model.named_parameters():
+            p.fake_device = torch.device('cuda:0')
+
+        # Need to set all the requires_grad of tensors to False, because fake_tensor with CUDA device
+        # doesn't quite work well with aot_autograd right now due to some logic fails
+        # the check in call getDeviceGuardImpl in InputMetadata.
+        x.requires_grad = False
+        for n, p in model.named_parameters():
+            p.requires_grad = False
+
```
Comment (SherlockNoMad) [inline test/export/test_export.py:1607]: I wonder if we can have a better UX for this workflow. 

I imagine something like this 

```
with fake_mode, torch.no_grad():
    x = torch.rand(2, device='cuda:0') # this sets x.fake_device under the hood
    model = Model()
    model.submodule1.to("cuda:0") # parameters are fakified and placed on "cuda:0"
    model.submodule2.to("cuda:1")
```

Comment (SherlockNoMad) [reply test/export/test_export.py:1607]: This mimics the normal model construction. 

In such case, use just need to add `with fake_mode`, and no other changes are needed.

**Review by SherlockNoMad — APPROVED** [review]
lgtm, and I wonder if we can make the UX better.

File: test/export/test_export.py (hunk 4 above)
Comment (ydwu4) [reply test/export/test_export.py:1607]: Let me check whether the UX can work. with torch.no_grad() doesn't set all the param/buffer's requires_grad to False somehow. Haven't checked the to("cuda:0") thing.

**Comment by SherlockNoMad** [comment]
Second though, to make this really useable, user shouldn't need to manipulate x.fake_device. User should write x.to("real_device"), and under fakemode, we handle the fake device placement for them.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing an internal ExecuTorch test' -c ghfirst

The error is as follows:

```
======================================================================
FAIL: test_asr_encoders (pye.model_inventory.asr_models.milan_dictation.test_xnnpack_asr.TestXnnpackPartitioner)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/model_inventory/asr_models/milan_dictation/test_xnnpack_asr.py", line 28, in test_asr_encoders
    exported_program = gen.get_exported_program(ModelVariant.FP32, methods[0])
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/lib/EagerModelBase.py", line 164, in wrapper
    return deepcopy(fn(*args, **kwargs))
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/model_inventory/asr_models/milan_dictation/MilanDictationModel.py", line 277, in get_exported_program
    exported_program = exir.capture(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/executorch/exir/capture/_capture.py", line 171, in capture
    ep = export(f, args, constraints=constraints)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 549, in export
    return _export(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 562, in wrapper
    return fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 615, in _export
    fake_args, fake_kwargs, fake_params_buffers, dynamo_fake_mode = _convert_input_to_fake(gm_torch_level, args, kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 412, in _convert_input_to_fake
    assert fake_mode is not None, "Cannot find fake_mode attatched to the graph's placeholders."
AssertionError: Cannot find fake_mode attatched to the graph's placeholders.

----------------------------------------------------------------------
Ran 1 test in 129.417s

FAILED (failures=1)
```

**Comment by ydwu4** [comment]
Internal tests have a model that takes no inputs, this makes _convert_to_fake_input unhappy. We detect this case and set the fake_mode to None in this case. Also added a new test for it.

**Comment by ydwu4** [comment]
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51414973).

**Review by angelayi — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
should this issue be closed now that #114009 was relanded?
